module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,280 @@
+func NewExecutionDataService(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+) *ExecutionDataService {
+	return &ExecutionDataService{&serializer{codec, compressor}, blobService, defaultMaxBlobSize}
+}
+
+func (s *ExecutionDataService) storeBatch(ctx context.Context, br *BlobReceiver) ([]network.Blob, error) {
+	var blobs []network.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob network.Blob
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, err
+}
+
+func (s *ExecutionDataService) storeBlobs(parent context.Context, br *BlobReceiver) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	var cids []cid.Cid
+
+	for {
+		batch, recvErr := s.storeBatch(ctx, br)
+
+		for _, blob := range batch {
+			cids = append(cids, blob.Cid())
+		}
+
+		if err := s.blobService.AddBlobs(ctx, batch); err != nil {
+			return nil, err
+		}
+
+		if recvErr != nil {
+			if recvErr != ErrClosedBlobChannel {
+				// this is an unexpected error, and should never occur
+				return nil, recvErr
+			}
+
+			break
+		}
+	}
+
+	return cids, nil
+}
+
+func (s *ExecutionDataService) addBlobs(ctx context.Context, v interface{}) ([]cid.Cid, error) {
+	bcw, br := IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			return
+		}
+
+		serializeErr = bcw.Flush()
+	}()
+
+	cids, recvErr := s.storeBlobs(ctx, br)
+
+	<-done
+
+	if recvErr != nil {
+		return nil, recvErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	cids, err := s.addBlobs(ctx, sd)
+
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	for {
+		if len(cids) == 1 {
+			return cids[0], nil
+		}
+
+		if cids, err = s.addBlobs(ctx, cids); err != nil {
+			return cid.Undef, fmt.Errorf("failed to add cid blobs: %w", err)
+		}
+	}
+}
+
+func (s *ExecutionDataService) retrieveBlobs(parent context.Context, bs *BlobSender, cids []cid.Cid) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids)
+	cachedBlobs := make(map[cid.Cid]network.Blob)
+
+	for _, c := range cids {
+		var err error
+		blob, ok := cachedBlobs[c]
+
+		if ok {
+			delete(cachedBlobs, c)
+		} else {
+			blob, err = s.findBlob(blobChan, c, cachedBlobs)
+
+			if err != nil {
+				return err
+			}
+
+			if blob == nil {
+				// the blob channel may be closed as a result of the context being canceled,
+				// in which case we should return the context error.
```

Comment:
Why would the context get canceled?

The context is only canceled when we [exit this function](https://github.com/onflow/flow-go/pull/1582/files#diff-06e9e7fa42b0b4d29eaa42ca5c75a865d70919934e07e97a75a02b63a9d10313R138), and here we are still within this function.

Reply:
The parent context which is passed in could be canceled.